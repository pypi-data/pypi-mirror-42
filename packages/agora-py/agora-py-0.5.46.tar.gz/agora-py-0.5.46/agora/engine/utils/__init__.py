"""
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Ontology Engineering Group
        http://www.oeg-upm.net/
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Copyright (C) 2016 Ontology Engineering Group.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

            http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
"""
import os
from threading import Lock

__author__ = 'Fernando Serena'


class Singleton(type):
    _lock = Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            id = kwargs.get('id', '')
            if id not in cls.instances:
                cls.instances[id] = super(Singleton, cls).__call__(*args, **kwargs)
            return cls.instances[id]


def prepare_store_path(base, path):
    if base and not os.path.exists(base):
        os.makedirs(base)
    if base or path:
        full_path = '/'.join(filter(lambda x: x, [base, path]))
        if not os.path.exists(full_path):
            os.makedirs(full_path)
        return full_path


class Semaphore(object):
    def __init__(self):
        self.value = -1

    def __enter__(self):
        self.value = 0

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.value = 1

    def isSet(self):
        return self.value == 1

    def set(self):
        self.value = 1


def get_immediate_subdirectories(a_dir):
    if not os.path.exists(a_dir):
        os.makedirs(a_dir)
    return [name for name in os.listdir(a_dir)
            if os.path.isdir(os.path.join(a_dir, name))]


class Wrapper(object):
    def __init__(self, object):
        self.__lock = Lock()
        self.__attr_locks = {}

        if isinstance(object, Wrapper):
            self.__object = object.__object
            self.__argret = object.__argret
            self.__attr_locks = object.__attr_locks
            self.__lock = object.__lock
        else:
            self.__object = object
            self.__argret = {}

    def decorate(self, f):
        def wrapper(*args, **kwargs):

            key = f.func_name + str(args)
            with self.__lock:
                if key not in self.__attr_locks:
                    self.__attr_locks[key] = Lock()

            with self.__attr_locks[key]:
                if key not in self.__argret:
                    ret = f(*args, **kwargs)

                    if kwargs:
                        return ret
                    if ret is not None:
                        self.__argret[key] = ret

                return self.__argret.get(key, None)

        return wrapper

    def __getattr__(self, item):

        if item.startswith('__'):
            return self.__object.__getattribute__(item)

        with self.__lock:
            if item not in self.__attr_locks:
                self.__attr_locks[item] = Lock()

        with self.__attr_locks[item]:
            if item not in self.__argret:
                attr = self.__object.__getattribute__(item)
                if hasattr(attr, '__call__'):
                    return self.decorate(attr)
                else:
                    self.__argret[item] = attr

        return self.__argret[item]


stopped = Semaphore()
