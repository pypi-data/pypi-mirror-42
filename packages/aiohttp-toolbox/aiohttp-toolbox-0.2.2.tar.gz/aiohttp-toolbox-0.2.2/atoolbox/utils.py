import json
import re
from typing import Any, Tuple, Type, TypeVar, Union

from aiohttp.web import Response
from aiohttp.web_exceptions import HTTPException
from pydantic import BaseModel, ValidationError, validate_model
from pydantic.fields import Shape

from .json_tools import pretty_lenient_json

JSON_CONTENT_TYPE = 'application/json'
IP_HEADER = 'X-Forwarded-For'
PROTO_HEADER = 'X-Forwarded-Proto'
URI_NOT_ALLOWED = re.compile(r'[^a-zA-Z0-9_\-/.]')
REMOVE_PORT = re.compile(r':\d{2,}$')
PydanticModel = TypeVar('PydanticModel', bound=BaseModel)

__all__ = (
    'raw_json_response',
    'json_response',
    'parse_request_json',
    'parse_request_json_ignore_missing',
    'parse_request_query',
    'get_ip',
    'request_root',
    'JsonErrors',
    'encrypt_json',
    'decrypt_json',
    'get_offset',
    'slugify',
    'remove_port',
    'RequestError',
)


def raw_json_response(json_str: Union[str, bytes, None], status_: int = 200):
    if isinstance(json_str, str):
        body = json_str.encode()
    elif isinstance(json_str, bytes):
        body = json_str
    elif json_str is None:
        body = b'null'
    else:
        raise TypeError(f'json_str must be bytes or str, not "{type(json_str)}')
    return Response(body=body + b'\n', status=status_, content_type=JSON_CONTENT_TYPE)


def json_response(*, status_=200, list_=None, headers_=None, **data):
    return Response(
        body=json.dumps(data if list_ is None else list_).encode() + b'\n',
        status=status_,
        content_type=JSON_CONTENT_TYPE,
        headers=headers_,
    )


async def parse_request_json(request, model: Type[PydanticModel], *, headers=None) -> PydanticModel:
    error_details = None
    try:
        data = await request.json()
    except ValueError:
        error_msg = 'Invalid JSON'
    else:
        try:
            return model.parse_obj(data)
        except ValidationError as e:
            error_msg = 'Invalid Data'
            error_details = e.errors()

    raise JsonErrors.HTTPBadRequest(message=error_msg, details=error_details, headers=headers)


def parse_request_query(request, model: Type[PydanticModel], *, headers=None) -> PydanticModel:
    data = {}
    for k in request.query:
        v = request.query.getall(k)
        f = model.__fields__.get(k)
        if len(v) > 1 or f and f.shape != Shape.SINGLETON:
            data[k] = v
        else:
            data[k] = v[0]

    try:
        return model(**data)
    except ValidationError as e:
        raise JsonErrors.HTTPBadRequest(message='Invalid query data', details=e.errors(), headers=headers)


async def parse_request_json_ignore_missing(
    request, model: Type[PydanticModel], *, headers=None
) -> Tuple[PydanticModel, dict]:
    try:
        raw_data = await request.json()
    except ValueError:
        raise JsonErrors.HTTPBadRequest(message='Invalid JSON', headers=headers)
    if not isinstance(raw_data, dict):
        raise JsonErrors.HTTPBadRequest(message='data not a dictionary', headers=headers)

    data, e = validate_model(model, raw_data, raise_exc=False)
    if e:
        errors = [e for e in e.errors() if not (e['type'] == 'value_error.missing' and len(e['loc']) == 1)]
        if errors:
            raise JsonErrors.HTTPBadRequest(message='Invalid Data', details=errors, headers=headers)

    return model.construct(**data), raw_data


def get_ip(request):
    ips = request.headers.get(IP_HEADER)
    if ips:
        return ips.split(',', 1)[0].strip(' ')
    else:
        return request.remote


def request_root(request):
    # request.url.scheme doesn't work as https is already terminated
    scheme = request.headers.get(PROTO_HEADER) or 'http'
    return f'{scheme}://{request.host}'


class JsonErrors:
    class _HTTPExceptionJson(HTTPException):
        custom_reason = None

        def __init__(self, message, *, details=None, headers=None):
            self.message = message
            self.details = details
            data = {'message': message}
            if details:
                data['details'] = details
            super().__init__(
                text=pretty_lenient_json(data),
                content_type=JSON_CONTENT_TYPE,
                headers=headers,
                reason=self.custom_reason,
            )

        def __repr__(self) -> str:
            return f'{super().__repr__()}, {self.status}: {self.message}'

        def __str__(self) -> str:
            return repr(self)

    class HTTPAccepted(_HTTPExceptionJson):
        status_code = 202

    class HTTPBadRequest(_HTTPExceptionJson):
        status_code = 400

    class HTTPUnauthorized(_HTTPExceptionJson):
        status_code = 401

    class HTTPPaymentRequired(_HTTPExceptionJson):
        status_code = 402

    class HTTPForbidden(_HTTPExceptionJson):
        status_code = 403

    class HTTPNotFound(_HTTPExceptionJson):
        status_code = 404

    class HTTPMethodNotAllowed(_HTTPExceptionJson):
        status_code = 405

        def __init__(self, message, allowed_methods, *, headers=None):
            headers = headers or {}
            headers.setdefault('Allow', ','.join(allowed_methods))
            super().__init__(message, details={'allowed_methods': allowed_methods}, headers=headers)

    class HTTPConflict(_HTTPExceptionJson):
        status_code = 409

    class HTTP470(_HTTPExceptionJson):
        status_code = 470
        custom_reason = 'Invalid user input'


def encrypt_json(app, data: Any) -> str:
    return app['auth_fernet'].encrypt(json.dumps(data).encode()).decode()


def decrypt_json(app, token: bytes, *, ttl: int = None, headers=None) -> Any:
    from cryptography.fernet import InvalidToken

    try:
        return json.loads(app['auth_fernet'].decrypt(token, ttl=ttl).decode())
    except InvalidToken:
        raise JsonErrors.HTTPBadRequest(message='invalid token', headers=headers)


def get_offset(request, paginate_by=100):
    page = request.query.get('page')
    if not page:
        return 0

    try:
        p = int(page)
        if p < 1:
            raise ValueError()
    except ValueError:
        raise JsonErrors.HTTPBadRequest(f"invalid page '{page}'")
    else:
        return (p - 1) * paginate_by


def slugify(title):
    name = title.replace(' ', '-').lower()
    name = URI_NOT_ALLOWED.sub('', name)
    name = re.sub('-{2,}', '-', name)
    return name.strip('_-')


def remove_port(host):
    return REMOVE_PORT.sub('', host)


class RequestError(RuntimeError):
    def __init__(self, status, url, *, text: str = None):
        self.status = status
        self.url = url
        self.text = text

    def __str__(self):
        return f'response {self.status} from "{self.url}"' + (f':\n{self.text[:400]}' if self.text else '')

    def json(self):
        return json.loads(self.text)

    def extra(self):
        return self.text
