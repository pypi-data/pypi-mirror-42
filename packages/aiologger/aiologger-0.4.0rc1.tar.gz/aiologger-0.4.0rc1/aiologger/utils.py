from typing import Callable


class classproperty:
    def __init__(self, func):
        self._func = func

    def __get__(self, obj, owner):
        return self._func(owner)


class CallableWrapper:
    def __init__(self, func: Callable) -> None:
        self.func = func

    def __call__(self, *args, **kwargs):
        return self.func(*args, **kwargs)
