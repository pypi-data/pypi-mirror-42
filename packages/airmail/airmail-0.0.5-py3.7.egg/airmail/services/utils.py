import os, sys
from pathlib import Path
import click
import yaml

def err_log(msg, *args):
    if args:
        msg %= args
    click.echo(click.style('Error: ', fg='red', bold=True) + msg)

# Read a yaml file given a file path. Shou
def read_yml(filepath):
    target_file = Path(filepath)

    if target_file.is_file():
        with open(target_file) as yml_file:
            return yaml.load(yml_file)
    else:
        click.echo(click.style('Error: ', fg='red', bold=True) +
            'Could not read the .yml at ' +
            click.style(filepath, bold=True))
        sys.exit()

# Determines the path to files inside the
# project. (i.e. the `data` directory)
def determine_project_path():
    """Borrowed from wxglade.py"""
    try:
        root = __file__
        if os.path.islink (root):
            root = os.path.realpath (root)
        return os.path.dirname (os.path.abspath (root))
    except:
        print("I'm sorry, but something is wrong.")
        print("There is no __file__ variable. Please contact the author.")
        sys.exit ()
