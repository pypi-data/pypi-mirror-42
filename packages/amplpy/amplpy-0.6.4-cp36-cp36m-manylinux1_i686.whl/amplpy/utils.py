# -*- coding: utf-8 -*-
from __future__ import print_function, absolute_import, division
from builtins import map, range, object, zip, sorted
from past.builtins import basestring, unicode
from sys import version_info

from . import amplpython
from .base import BaseClass


def lock_and_call(call, lock):
    lock.acquire()
    try:
        result = call()
    except Exception:
        lock.release()
        raise
    else:
        lock.release()
        return result


class Tuple(BaseClass):
    def __init__(self, values=None, **kwargs):
        if values is not None:
            values = Utils.convToList(values)
            va = Utils.toVariantArray(values)
            self._impl = amplpython.Tuple.Factory(va, len(values))
        else:
            self._impl = kwargs.get('_impl', None)

    def toTuple(self):
        return tuple(
            Utils.castVariant(self._impl.getIndex(i))
            for i in range(self._impl.size())
        )

    def toPyObj(self):
        size = self._impl.size()
        if size == 1:
            return Utils.castVariant(self._impl.getIndex(0))
        else:
            return self.toTuple()

    @classmethod
    def fromTupleRef(cls, tupleRef):
        return cls(_impl=amplpython.Tuple.Factory(tupleRef))


class Utils(object):
    @staticmethod
    def toVariantArray(lst):
        va = amplpython.VariantArray(len(lst))
        for i in range(len(lst)):
            if isinstance(lst[i], unicode):
                va[i] = amplpython.Variant(str(lst[i]))
                # FIXME: This is just a workaround for issue amplapi#332
                # The real fix requires a new release of amplapi
            else:
                va[i] = amplpython.Variant(lst[i])
        return va

    @staticmethod
    def toTupleArray(lst):
        ta = amplpython.TupleArray(len(lst))
        for i in range(len(lst)):
            ta[i] = Tuple(lst[i])._impl
        return ta

    @staticmethod
    def castVariant(variant):
        if variant.type() == amplpython.NUMERIC:
            return variant.dbl()
        elif variant.type() == amplpython.STRING:
            return variant.str()

    @staticmethod
    def castStringArray(sarray):
        return [sarray.getIndex(i) for i in range(sarray.size())]

    @staticmethod
    def castVariantRef(variantref):
        return Utils.castVariant(amplpython.Variant(variantref))

    @staticmethod
    def convToList(value):
        if isinstance(value, list):
            return value
        elif isinstance(value, tuple):
            return list(value)
        else:
            return [value]


def multidict(d):
    try:
        ncols = min(len(d[k]) for k in d)
    except TypeError:
        raise TypeError
    return [list(d.keys())] + [{k: d[k][i] for k in d} for i in range(ncols)]
