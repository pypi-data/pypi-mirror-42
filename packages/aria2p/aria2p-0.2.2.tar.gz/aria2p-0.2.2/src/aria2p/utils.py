"""
Utils module.

This module contains simple utility functions.
"""


def human_readable_timedelta(value):
    pieces = []

    if value.days:
        pieces.append(f"{value.days}d")

    seconds = value.seconds

    if seconds >= 3600:
        hours = int(seconds / 3600)
        pieces.append(f"{hours}h")
        seconds -= hours * 3600

    if seconds >= 60:
        minutes = int(seconds / 60)
        pieces.append(f"{minutes}m")
        seconds -= minutes * 60

    if seconds > 0 or not pieces:
        pieces.append(f"{seconds}s")

    return "".join(pieces)


def human_readable_bytes(value, digits=2, delim="", postfix=""):
    unit = "B"
    for u in ("KiB", "MiB", "GiB", "TiB"):
        if value > 1000:
            value /= 1024
            unit = u
        else:
            break
    return f"{value:.{digits}f}" + delim + unit + postfix


def bool_or_value(value):
    if value == "true":
        return True
    elif value == "false":
        return False
    else:
        return value


def bool_to_str(value):
    if value is True:
        return "true"
    elif value is False:
        return "false"
    else:
        return value
