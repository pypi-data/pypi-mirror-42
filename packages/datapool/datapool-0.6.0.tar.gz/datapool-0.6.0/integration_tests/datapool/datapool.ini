[db]
connection_string = sqlite+pysqlite:////Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz/.simple.db

[backup_landing_zone]
folder = 

[r]
extension = .r
executable = /usr/local/bin/R

[matlab]
extension = .m
executable = /Applications/MATLAB_R2016a.app/bin/matlab

[julia]
extension = .jl
version = 0.5.2
executable = /usr/local/bin/julia

[python]
extension = .py
executable = /Users/uweschmitt/Projects/eawag_datapool/venv/bin/python3

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[landing_zone]
folder = /Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz

