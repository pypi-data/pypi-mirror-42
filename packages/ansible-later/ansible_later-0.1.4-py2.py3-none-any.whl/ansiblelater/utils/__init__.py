from __future__ import print_function

import importlib
import logging
import os
import subprocess
import sys
import re

from distutils.version import LooseVersion

try:
    import ConfigParser as configparser
except ImportError:
    import configparser

try:
    from ansible.utils.color import stringc
except ImportError:
    from ansible.color import stringc

# from yamlhelper import *


def abort(message, file=sys.stderr):
    print(stringc("FATAL: %s" % message, 'red'), file=file)
    sys.exit(1)


def error(message, file=sys.stderr):
    print(stringc("ERROR: %s" % message, 'red'), file=file)


def warn(message, settings, file=sys.stdout):
    if settings.log_level <= logging.WARNING:
        print(stringc("WARN: %s" % message, 'yellow'), file=file)


def info(message, settings, file=sys.stdout):
    if settings.log_level <= logging.INFO:
        print(stringc("INFO: %s" % message, 'green'), file=file)


def count_spaces(c_string):
    leading_spaces = 0
    trailing_spaces = 0

    for i, e in enumerate(c_string):
        if not e.isspace():
            break
        leading_spaces += 1

    for i, e in reversed(list(enumerate(c_string))):
        if not e.isspace():
            break
        trailing_spaces += 1

    return((leading_spaces, trailing_spaces))


def get_property(prop):
    currentdir = os.path.dirname(os.path.realpath(__file__))
    parentdir = os.path.dirname(currentdir)
    result = re.search(
        r'{}\s*=\s*[\'"]([^\'"]*)[\'"]'.format(prop),
        open(os.path.join(parentdir, '__init__.py')).read())
    return result.group(1)


def standards_latest(standards):
    return max([standard.version for standard in standards if standard.version] or ["0.1"],
               key=LooseVersion)


def lines_ranges(lines_spec):
    if not lines_spec:
        return None
    result = []
    for interval in lines_spec.split(","):
        (start, end) = interval.split("-")
        result.append(range(int(start), int(end) + 1))
    return result


def is_line_in_ranges(line, ranges):
    return not ranges or any([line in r for r in ranges])


def read_standards(settings):
    if not settings.rulesdir:
        abort("Standards directory is not set on command line or in configuration file - aborting")
    sys.path.append(os.path.abspath(os.path.expanduser(settings.rulesdir)))
    try:
        standards = importlib.import_module('standards')
    except ImportError as e:
        abort("Could not import standards from directory %s: %s" % (settings.rulesdir, str(e)))
    return standards


def read_config(config_file):
    config = configparser.RawConfigParser({'standards': None})
    config.read(config_file)

    return Settings(config, config_file)


def execute(cmd):
    result = ExecuteResult()
    encoding = 'UTF-8'
    env = dict(os.environ)
    env['PYTHONIOENCODING'] = encoding
    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE,
                            stderr=subprocess.STDOUT, env=env)
    result.output = proc.communicate()[0].decode(encoding)
    result.rc = proc.returncode
    return result


class Settings(object):
    def __init__(self, config, config_file):
        self.rulesdir = None
        self.custom_modules = []
        self.log_level = None
        self.standards_filter = None

        if config.has_section('rules'):
            self.rulesdir = config.get('rules', 'standards')
        if config.has_section('ansible'):
            modules = config.get('ansible', 'custom_modules')
            self.custom_modules = [x.strip() for x in modules.split(',')]

        self.configfile = config_file


class ExecuteResult(object):
    pass
