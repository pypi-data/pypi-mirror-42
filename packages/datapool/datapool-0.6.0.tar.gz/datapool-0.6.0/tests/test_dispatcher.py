# encoding: utf-8
from __future__ import absolute_import, division, print_function

import os
import shutil
import time

from datapool.database import dump_db, setup_db
from datapool.dispatcher import Dispatcher, log_unexpected_exceptions
from datapool.utils import find_executable


def test_log_unexpected_exceptions(regtest_logger):
    @log_unexpected_exceptions
    def by_zero(a, b, c):
        return a + b / c

    with regtest_logger():
        by_zero(1, 2, 0)
        by_zero("1", 2, c=0)
        by_zero({1}, b="2", c=0)


def _run_dispatcher(config, regtest_logger, lz, paths):
    setup_db(config.db)
    with regtest_logger() as logger:
        dispatcher = Dispatcher(config, time_provider=lambda: "NOW", info=logger.info)
        for path in paths:
            logger.info("dispatch: {}".format(path))
            for message in dispatcher.dispatch(path, time.time()):
                logger.info("message: {}".format(message))


def test_dispatcher_ok(config, regtest, regtest_logger, lz):
    paths = (
        "sites/example_site/site.yaml",
        "data/parameters.yaml",
        "data/FloDar/source_type.yaml",
        "data/FloDar/FloDar_python/source.yaml",
        "data/FloDar/FloDar_python/conversion.py",
        "data/FloDar/FloDar_python/raw_data/data-001.raw",
    )

    config.python.executable = find_executable("python")
    _run_dispatcher(config, regtest_logger, lz, paths)
    dump_db(config.db, file=regtest, max_rows=25)
    for path in lz.list_all_files():
        print(path, file=regtest)

    print("\nfiles backup landing zone:".upper(), file=regtest)
    root_backup = config.backup_landing_zone.folder
    for dirname, dirs, files in os.walk(root_backup):
        dirname = os.path.relpath(dirname, root_backup)
        for path in files:
            print("  ", os.path.join(dirname, path), file=regtest)


def test_dispatcher_corrupt_raw_file(config, regtest, regtest_logger, lz, tmpdir):

    setup_db(config.db)
    dev_root = config.landing_zone.folder
    test_root = config.landing_zone.folder = tmpdir.strpath
    config.python.executable = find_executable("python")

    def one_run(run_index):

        paths = (
            "sites/example_site/site.yaml",
            "sites/example_site/images/1.JPG",
            "sites/example_site/images/2.JPG",
            "sites/example_site/images/3.JPG",
            "sites/example_site/site.yaml",
            "data/parameters.yaml",
            "data/FloDar/source_type.yaml",
            "data/FloDar/FloDar_python/source.yaml",
            "data/FloDar/FloDar_python/conversion.py",
            "data/FloDar/FloDar_python/raw_data/data-001.raw",
        )

        for p in paths:
            target = os.path.join(test_root, p)
            target_folder = os.path.dirname(target)
            if not os.path.exists(target_folder):
                os.makedirs(target_folder)
            shutil.copy(os.path.join(dev_root, p), target)

        assert target.endswith(
            ".raw"
        ), "someone broke this test by changing 'paths' above"

        duplicate_signals_diff_values = "\t".join(["13.11.2013 10:06"] + 7 * ["999"])

        lines = open(target, encoding="latin-1").readlines()
        with open(target, "w", encoding="latin-1") as fh:
            fh.writelines(lines[0:2])
            fh.writelines(lines[1])
            if run_index == 1:
                fh.write(duplicate_signals_diff_values)

        with regtest_logger() as logger:
            dispatcher = Dispatcher(
                config, time_provider=lambda: "NOW", info=logger.info
            )
            for path in paths:
                for message in dispatcher.dispatch(path, time.time()):
                    logger.info("message: {}".format(message))

    one_run(0)
    dump_db(
        config.db,
        ["site", "parameter", "source", "source_type", "signal"],
        file=regtest,
        max_rows=25,
    )

    # another run with dupicate signal with different value:
    one_run(1)
    dump_db(config.db, ["signal"], file=regtest, max_rows=25)

    print("\nlanding zone:".upper(), file=regtest)
    for path in lz.list_all_files():
        print("  ", path, file=regtest)

    print("\nfiles backup landing zone:".upper(), file=regtest)
    root_backup = config.backup_landing_zone.folder
    for dirname, dirs, files in os.walk(root_backup):
        dirname = os.path.relpath(dirname, root_backup)
        for path in files:
            print("  ", os.path.join(dirname, path), file=regtest)


def test_dispatcher_no_python(config, regtest_logger, lz):
    # we don't configure r here to trigger an error
    paths = (
        "data/FloDar/FloDar_python/conversion.py",
        "data/FloDar/FloDar_python/raw_data/data-001.raw",
    )

    _run_dispatcher(config, regtest_logger, lz, paths)


def test_dispatcher_no_raw(config, regtest_logger, lz):
    lz.remove_file("data/FloDar/FloDar_python/raw_data/data-001.raw")
    paths = ("data/FloDar/FloDar_python/conversion.py",)
    _run_dispatcher(config, regtest_logger, lz, paths)


def test_dispatcher_no_script(config, regtest_logger, lz):
    lz.remove_file("data/FloDar/FloDar_python/conversion.py")
    paths = ("data/FloDar/FloDar_python/raw_data/data-001.raw",)
    _run_dispatcher(config, regtest_logger, lz, paths)


def test_dispatcher_multiple_scripts(config, regtest_logger, lz):
    lz.overwrite_file(
        "data/FloDar/FloDar_python/conversion.jl",
        lz.p("data/FloDar/FloDar_python/conversion.py"),
    )
    paths = ("data/FloDar/FloDar_python/raw_data/data-001.raw",)
    _run_dispatcher(config, regtest_logger, lz, paths)


def test_dispatcher_wrong_yaml_name(config, regtest_logger, lz):
    paths = ("data/invalid_name.yaml",)
    for p in paths:
        open(lz.p(p), "w").close()
    try:
        _run_dispatcher(config, regtest_logger, lz, paths)
    finally:
        for p in paths:
            os.remove(lz.p(p))


def test_dispatcher_invalid_yaml(config, regtest_logger, data_path, lz):
    lz.overwrite_file("data/parameters.yaml", data_path("yamls/empty.yaml"))
    paths = ("data/parameters.yaml",)
    _run_dispatcher(config, regtest_logger, lz, paths)


def test_dispatcher_unsupported_file_extension(config, regtest_logger, lz):
    # we don't configure r here to trigger an error
    paths = ("new_file.txt",)
    _run_dispatcher(config, regtest_logger, lz, paths)
