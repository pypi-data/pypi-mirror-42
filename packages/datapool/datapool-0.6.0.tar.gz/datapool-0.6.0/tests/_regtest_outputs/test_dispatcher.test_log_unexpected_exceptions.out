ERROR    eawag.datapool  unexpected error 'division by zero' when calling by_zero(1, 2, 0)
ERROR    eawag.datapool  unexpected error 'division by zero' when calling by_zero('1', 2, c=0)
ERROR    eawag.datapool  unexpected error 'unsupported operand type(s) for /: 'str' and 'int'' when calling by_zero({1}, b='2', c=0)
